WebRoot/js/zimbraMail/share/view/dialog/ZmTwoFactorSetupDialog.js

```diff
@@ -362,18 +672,67 @@ function() {
 };
 
 ZmTwoFactorSetupDialog.disableTwoFactorAuth =
-function(params, dialog) {
+function(params, dialog, method) {
 	var command = new ZmCsfeCommand();
-	var jsonObj = {DisableTwoFactorAuthRequest : {_jsns:"urn:zimbraAccount"}};
-	var callback = ZmTwoFactorSetupDialog.disableTwoFactorAuthCallback.bind(window, params, dialog);
+	var jsonObj = {DisableTwoFactorAuthRequest : {_jsns:"urn:zimbraAccount", method:{_content : method}}};
+	var callback = ZmTwoFactorSetupDialog.disableTwoFactorAuthCallback.bind(window, params, dialog, method);
 	command.invoke({jsonObj: jsonObj, noAuthToken: true, asyncMode: true, callback: callback, serverUri:"/service/soap/"});
 };
 
 ZmTwoFactorSetupDialog.disableTwoFactorAuthCallback =
-function(params, dialog) {
-	dialog.popdown();
-	Dwt.setInnerHtml(params.twoStepAuthSpan, ZmMsg.twoStepStandardAuth);
-	Dwt.setInnerHtml(params.twoStepAuthLink, ZmMsg.twoStepAuthSetupLink);
-	Dwt.setDisplay(params.twoStepAuthCodesContainer, Dwt.DISPLAY_NONE);
-	appCtxt.set(ZmSetting.TWO_FACTOR_AUTH_ENABLED, false, false, false, true);
-};
\ No newline at end of file
+function(params, dialog, method) {
+	if (dialog) {
+		dialog.popdown();
+	}
+	var callback = ZmTwoFactorSetupDialog._GetInfoCallback.bind(this, params, true, method);
+	var command = new ZmCsfeCommand();
+	var jsonObj = {GetInfoRequest : {_jsns:"urn:zimbraAccount"}};
+	var result = command.invoke({jsonObj: jsonObj, noAuthToken: true, asyncMode: true, callback: callback, serverUri: "/service/soap/"});
+};
+
+/**
+ * Get account settings from a server because some attributes are calculated and set on the server.
+ * It is non-prototype function so that it can be called from ZmTwoFactorSetupDialog.disableTwoFactorAuthCallback.
+ *
+ * @param {object}   params       The object parameters which are passed at DisableTwoFactorAuthRequest
+ * @param {boolean}  disableTFA   Whether GetInfoRequest is called after DisableTwoFactorAuthRequest
+ * @param {string}   method       The name of a TFA method being configured
+ * @param {object}   result       The response object
+ */
+ZmTwoFactorSetupDialog._GetInfoCallback =
+function(params, disableTFA, method, result) {
+	var dialog = appCtxt.getMsgDialog();
+	dialog.setMessage(ZmMsg.twoFactorAuthGetInfoFailed, DwtMessageDialog.WARNING_STYLE);
+	if (!result || result.isException()) {
+		dialog.popup();
+		return;
+	}
+	try {
+		var response = result.getResponse().Body.GetInfoResponse;
```

Comment:
we are already initializing all settings based on GetInfo call at startup check settings.setUserSettings
so do we need another call here?

Reply:
yes, regarding zimbraTwoFactorAuthMethodEnabled, zimbraPrefPrimaryTwoFactorAuthMethod and zimbraPrefPasswordRecoveryAddressStatus, they are changed depending on the current configuration.
The logic is a bit complicated for backward compatibility and for consistency between TFA setting and recovery account setting.
The logic may be updated for some reason. (e.g. new method is introduced in the future, any issue is found, etc.)
It is better to get the updated settings using GetInfoRequest/Response for consistency between UI and server, and to avoid duplicate work on server and UI changes.